lms/djangoapps/grades/course_grade.py

```diff
@@ -55,12 +55,22 @@ def subsection_grade(self, subsection_key):
         """
         Returns the subsection grade for the given subsection usage key.
 
-        Note: does NOT check whether the user has access to the subsection.
-        Assumes that if a grade exists, the user has access to it.  If the
-        grade doesn't exist then either the user does not have access to
-        it or hasn't attempted any problems in the subsection.
-        """
-        return self._get_subsection_grade(self.course_data.effective_structure[subsection_key])
+        This will always succed if there is a subsection with the given key,
+        regardless of the user's access;
+        it is up to the caller to avoid leaking information by only calling
+        this with appropriate values.
```

Comment:
I think it might be good to leave in this message: 

> [it's safe to assume that] If the        grade doesn't exist then either the user does not have access to        it or hasn't attempted any problems in the subsection.

Reply:
I think what I wrote in the docs could be misleading, so I changed to be clearer (IMO); does that address your comment? Otherwise I'm not sure what you mean.

Reply:
@kewne Sorry for not being clear.

I still think that these deleted lines:

> Assumes that if a grade exists, the user has access to it.  If the
> grade doesn't exist then either the user does not have access to
> it or hasn't attempted any problems in the subsection.

Provided a clearer explanation than:

> it is up to the caller to avoid leaking information by only calling
> this with appropriate values.

With the second, it says "there's a risk of leaking information" but doesn't really say what information could be leaked, or what "appropriate values" are.

In the first case it's pretty clear that most of the time there won't be any problems, and the only edge case would be if a user had access to a section, got a grade, then lost that access, and it implies that we don't really need to worry about that.

If you just replace the part that says "it is up to the caller to avoid leaking information by only calling this with appropriate values." with "We assume that if a grade exists, the user has access to it.  If the grade doesn't exist then either the user does not have access to it or hasn't attempted any problems in the subsection." then I think it's nice and clear.

What you have isn't bad, I just think it's slightly less useful. Sorry, I didn't mean to hold up the PR on this minor detail. I'm fine with merging it either way.

Reply:
@bradenmacdonald no worries, I wasn't getting what you meant; I've pushed another change to the docstring which, although it's not your suggestion, is hopefully clearer.

IMO "grade doesn't exist" suggests `None` is a possible return value, which isn't true in either the old or new code.
It could have meant "section not attempted" but I don't think that should be documented here because it's already part of the `Grade` model.

Reply:
Thanks, that seems even better then.